helix-core/src/test/java/org/apache/helix/integration/rebalancer/WagedRebalancer/TestWagedLoadedCluster.java

```diff
@@ -0,0 +1,277 @@
+import org.apache.helix.PropertyKey;
+import org.apache.helix.common.ZkTestBase;
+import org.apache.helix.controller.rebalancer.waged.AssignmentMetadataStore;
+import org.apache.helix.integration.manager.ClusterControllerManager;
+import org.apache.helix.integration.manager.MockParticipantManager;
+import org.apache.helix.manager.zk.ZKHelixDataAccessor;
+import org.apache.helix.manager.zk.ZkBucketDataAccessor;
+import org.apache.helix.model.BuiltInStateModelDefinitions;
+import org.apache.helix.model.ClusterConfig;
+import org.apache.helix.model.CurrentState;
+import org.apache.helix.model.Message;
+import org.apache.helix.model.ResourceAssignment;
+import org.apache.helix.model.ResourceConfig;
+import org.apache.helix.participant.StateMachineEngine;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+/**
+ * This test case is specially targeting for n - n+1 issue.
+ *
+ * Initially, all the partitions are equal size and equal weight
+ * from both CU, DISK.
+ * All the nodes are equally loaded.
+ * The test case will do the following:
+ *     For one resource, we will double its CU weight.
+ *     The rebalancer will be triggered.
+ *
+ * We have a monitoring thread which is constantly monitoring the instance capacity.
+ * - It looks at current state resource assignment + pending messages
+ * - it  has ASSERT in place to make sure we NEVER cross instance capacity (CU)
+ */
+public class TestWagedLoadedCluster extends ZkTestBase {
+  protected final int NUM_NODE = 6;
+  protected static final int START_PORT = 13000;
+  protected static final int PARTITIONS = 10;
+
+  protected static final String CLASS_NAME = TestWagedLoadedCluster.class.getSimpleName();
+  protected static final String CLUSTER_NAME = CLUSTER_PREFIX + "_" + CLASS_NAME;
+  protected ClusterControllerManager _controller;
+  protected AssignmentMetadataStore _assignmentMetadataStore;
+  List<MockParticipantManager> _participants = new ArrayList<>();
+  List<String> _nodes = new ArrayList<>();
+  private final Set<String> _allDBs = new HashSet<>();
+
+  private CountDownLatch _completedTest = new CountDownLatch(1);
+  private CountDownLatch _weightUpdatedLatch = new CountDownLatch(1); // when 0, weight is updated
+  private final Map<String, Integer> _defaultInstanceCapacity =
+      ImmutableMap.of("CU", 50, "DISK", 50);
+
+  private final Map<String, Integer> _defaultPartitionWeight =
+      ImmutableMap.of("CU", 10, "DISK", 10);
+
+  private final Map<String, Integer> _newPartitionWeight =
+      ImmutableMap.of("CU", 20, "DISK", 10);
+  private final int DEFAULT_DELAY = 500; // 0.5 second
+  private static final Logger LOG = LoggerFactory.getLogger(CLASS_NAME);
+
+  @BeforeClass
+  public void beforeClass() throws Exception {
+    LOG.info("START " + CLASS_NAME + " at " + new Date(System.currentTimeMillis()));
```

Comment:
Highly recommend to have a WAGED test base with basic beforeClass, afterClass and other setups. Otherwise, there are a lot of duplicated code in setup across several tests.

For this PR, let's keep as it for bug fixing. For future ones, let's do the WAGED base class thing.

Reply:
I wil do as a follow up change. thanks